torch/csrc/cuda/Event.cpp

```diff
@@ -0,0 +1,179 @@
+#include <torch/csrc/cuda/Event.h>
+#include <torch/csrc/cuda/Stream.h>
+
+#include <torch/csrc/THP.h>
+#include <torch/csrc/cuda/Module.h>
+
+#include <c10/cuda/CUDAGuard.h>
+
+#include <structmember.h>
+#include <cuda_runtime_api.h>
+
+PyObject *THCPEventClass = nullptr;
+
+static PyObject * THCPEvent_pynew(
+    PyTypeObject *type, PyObject *args, PyObject *kwargs) {
+  HANDLE_TH_ERRORS
+
+  int current_device;
+  THCudaCheck(cudaGetDevice(&current_device));
+
+  unsigned char enable_timing = 0;
+  unsigned char blocking = 0;
+  unsigned char interprocess = 0;
+  const char * handle_bytes = nullptr;
+  int handle_size = 0;
+
+  // cannot use bool 'p' and bytearray 'Y' as they are not available in Python 2
+  static char *kwlist[] =
+    {"enable_timing", "blocking", "interprocess", "_handle", nullptr};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|bbbs#", kwlist,
+      &enable_timing, &blocking, &interprocess, &handle_bytes, &handle_size)) {
+    return nullptr;
+  }
+
+  THPObjectPtr ptr(type->tp_alloc(type, 0));
+  if (!ptr) {
+    return nullptr;
+  }
+
+  THCPEvent* self = (THCPEvent *)ptr.get();
+  if (handle_bytes) {
+    AT_CHECK(handle_size == sizeof(cudaIpcEventHandle_t),
+      "cudaIpcEventHandle_t expects byte array of size ",
+      sizeof(cudaIpcEventHandle_t), ", but got ", handle_size);
+    // no need to release the handle byte array as it is automatically managed
+    // by the corresponding THCPEvent python object.
+    // see https://docs.python.org/3/c-api/arg.html#strings-and-buffers
+    new (&self->cuda_event) at::cuda::CUDAEvent(
+      (const cudaIpcEventHandle_t *) handle_bytes);
```

Comment:
I'm not sure if this cast violates the strict aliasing rule in C++. It might not because `handle_bytes` is a `char*`, but I'm not sure. It may still violate the rule because of the underlying type of the object in the Python API. 

To be safe, I think it's slightly preferable to copy into a `cudaIpcEventHandle_t` here. Something like:

```c++
cudaIpcEventHandle_t handle;
std::mempcy(&handle, handle_bytes, sizeof(handle));
new (&self->cuda_event) at::cuda::CUDAEvent(&handle);
```